chore(blog): Blog on rustyscript implementation in superposition (#1064)
juspay/superposition: docs/blog/2026-06-22-rustyscript.md

---
slug: js-sandbox-in-rust

> title:  Rust JS Sandbox - Running Untrusted JavaScript Inside Superposition

Reviewer: Can we change untrusted to user defined ?

Author: The meaning is the same - we want to isolate and execute the code.

> title:  Rust JS Sandbox - Running Untrusted JavaScript Inside Superposition

Reviewer 2: Double space after `title:`.

```suggestion
title: Rust JS Sandbox - Running Untrusted JavaScript Inside Superposition
```

> description: The design and implementation of validation and value compute functions - running user defined JS functions inside rust that is seamless and isolated

Reviewer 2: `user defined` should be hyphenated, `rust` should be capitalised, and the `that is seamless and isolated` clause is grammatically ambiguous (what does `that` refer to?).

```suggestion
description: Designing and implementing validation and value-compute functions: how we run user-defined JS inside Rust, seamlessly and in isolation.
```

tags: [technical deepdive, rust, js, functions]
---

> 

Reviewer: Can we add a section for what all we sandboxed ?

Author: We sandboxed functions only is my understanding, can you give an example of what this section would be

Reviewer 2: I think @sauraww means to ask what all capabilities are available for functions to execute (basically the rest of the world is denied right) ?

Is that right @sauraww ?

Reviewer: yes @knutties

> TL;DR - We used RustyScript to sandbox user submitted JS code inside our rust application

Reviewer 2: ```suggestion
TL;DR - We use RustyScript to execute user submitted JS code in a sandboxed fashion inside our rust application.
```

Author: Will replace this with `We use RustyScript to sandbox and execute user submitted JS code inside our rust application`

> TL;DR - We used RustyScript to sandbox user submitted JS code inside our rust application

Reviewer 2: `user-submitted`, `Rust`, and the crate is consistently `Rustyscript`/`rustyscript` (not `RustyScript`) in its own docs — pick one form and apply globally.

```suggestion
TL;DR - We used Rustyscript to sandbox user-submitted JS code inside our Rust application
```

> [Superposition](https://github.com/juspay/superposition) is a context-aware configuration management system that ensures correct, safe and reliable config changes. One of the ways we've achieved this goal is with jsonschema types for all configuration values with a default value. We call this concept [Default Configs](/docs/basic-concepts/context-aware-config/default-config). We soon realized an issue - though the type is right; that does not guarantee the correctness of the actual value. 

Reviewer 2: A few issues:
- Missing Oxford comma (`correct, safe, and reliable`) — used inconsistently elsewhere in the post.
- `One of the ways we've achieved this goal is with … with …` double-`with` reads awkwardly.
- The semicolon in `though the type is right; that does not guarantee …` is incorrect — should be a comma or em-dash.
- Trailing whitespace.

```suggestion
[Superposition](https://github.com/juspay/superposition) is a context-aware configuration management system that ensures correct, safe, and reliable config changes. One of the ways we achieve this is by requiring a JSON Schema and a default for every configuration value. We call this concept [Default Configs](/docs/basic-concepts/context-aware-config/default-config). We soon realized an issue — though the type is right, that does not guarantee the correctness of the actual value.
```

> 

Reviewer: Can we add example of URL to make the argument stronger ?

Author: Its in the example below

# Where types aren't enough

Take the following configuration:

```json
{
    "image_url": "https://example.com/image/cat.png"
}
```

> The most strict JSON schema for the value of image_url would be:

Reviewer 2: `most strict` → `strictest`.

```suggestion
The strictest JSON schema for the value of image_url would be:
```

> ```

Reviewer 2: Missing language tag on the fence — should be ` ```json ` so the block renders with syntax highlighting (matching the JSON block above). Same issue on the JS pseudo-signature at L35.

{
  "type": "string",
  "pattern": "^https?:\\/\\/(www\\.)?[-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6}\\b([-a-zA-Z0-9()@:%_\\+.~#?&//=]*)$"
}
```
This schema validates the type and shape and that is one part of correctness, but we're missing the most critical aspect - does the URL exist? Does it return what we're expecting?

# User Defined Validation Functions!

> We came up with functions - a way for any end user to write validation functions and link it to a Default Config or [Dimension](/docs/basic-concepts/context-aware-config/dimensions) to ensure the correctness of that value (should they want to do that). These functions are written in our frontend monaco editor and saved to a database with some other information. The function definition looks like this:

Reviewer 2: - `functions … validation functions` is repetitive; naming the feature distinctly reads better.
- `link it` — antecedent (`functions`) is plural, so `link them`.

```suggestion
We introduced *Functions*, a way for any end user to write validation logic and link them to a Default Config or [Dimension](/docs/basic-concepts/context-aware-config/dimensions) to ensure the correctness of that value (should they want to do that). These functions are written in our frontend monaco editor and saved to a database with some other information. The function definition looks like this:
```

> ```

Reviewer 2: Same as L23 — add a language tag (e.g. ` ```typescript ` since the `-> boolean` return annotation isn't valid JS).

Author: This is not valid JS syntax, just a formal representation so readers understand the shape of the function

Reviewer 2: We can write a working code segment right ?  Would readers understand that it is not valid JS?  Or we should call it pseudo-code explicitly for the sake of clarity.

Author: >We can write a working code segment right ? Would readers understand that it is not valid JS

Would it make sense for a just function? It should be obvious to readers that it is not valid JS, if not when reading definitely when they try to execute it :)

> async function execute(payload) -> boolean

Reviewer: this is wrong js syntax

```
async function execute(payload) {
  return true; // boolean
}
```

Author: This is a function definition, not JS code. Its an example

Reviewer 2: But example can be valid JS code right - just don't want to introduce custom code representations which an average reader might not be familiar with or can get confused.

We don't control readers - but we can afford to be clear from get go is what the broader spirit of the comment is.

Author: Just want to define the function inputs and outputs, the code representation would need a bit more cognitive effort to read this

```

> If the function returned false, the value trying to be set is invalid and Superposition would not allow it to be saved as a configuration change.

Reviewer 2: Tense mismatch (`returned` vs `is`); `the value trying to be set` reads awkwardly.

```suggestion
If the function returns false, the value being set is invalid and Superposition will not allow it to be saved as a configuration change.
```

# First implementation - NodeJS isolates

> Since we just needed a boolean return value, we wrapped the user code with Isolates and ran it as a nodeJS script. If the function returned false, the code forced the script to terminate with an custom exit code. When this was run through Rust's `std::process::Command`, we checked the exit code and deciphered the return value of the function. Users also asked for logs, so console.log would output to stdout, which we also read and returned to the user. Since this was NodeJS, any npm package like axios for HTTP requests we wanted to make available to the user could just be installed and used along with the script we we're running. 

Reviewer: Isolates makes it a little confusing.

Author: That is the technical term

> Since we just needed a boolean return value, we wrapped the user code with Isolates and ran it as a nodeJS script. If the function returned false, the code forced the script to terminate with an custom exit code. When this was run through Rust's `std::process::Command`, we checked the exit code and deciphered the return value of the function. Users also asked for logs, so console.log would output to stdout, which we also read and returned to the user. Since this was NodeJS, any npm package like axios for HTTP requests we wanted to make available to the user could just be installed and used along with the script we we're running. 

Reviewer: with a custom exit code

> Since we just needed a boolean return value, we wrapped the user code with Isolates and ran it as a nodeJS script. If the function returned false, the code forced the script to terminate with an custom exit code. When this was run through Rust's `std::process::Command`, we checked the exit code and deciphered the return value of the function. Users also asked for logs, so console.log would output to stdout, which we also read and returned to the user. Since this was NodeJS, any npm package like axios for HTTP requests we wanted to make available to the user could just be installed and used along with the script we we're running. 

Reviewer 2: Multiple fixes in one paragraph:
- `Isolates` — capitalisation reads as a proper noun; if you mean V8 Isolates or the `isolated-vm` npm package, name it explicitly.
- `nodeJS` / `NodeJS` → `Node.js` (canonical).
- `an custom exit code` → `a custom exit code`.
- `deciphered` is an odd choice — `interpreted` fits better.
- `we we're running` → `we were running`.
- The last sentence is hard to parse; splitting helps.

```suggestion
Since we just needed a boolean return value, we wrapped the user code with isolates and ran it as a Node.js script. If the function returned false, the code forced the script to terminate with a custom exit code. When this was run through Rust's `std::process::Command`, we checked the exit code and interpreted the return value of the function. Users also asked for logs, so console.log would output to stdout, which we also read and returned to the user. Because this ran under Node.js, users could install any npm package (e.g. axios for HTTP) and use it alongside the script we were running.
```

> 

Reviewer: `we were running`

When a function is being edited/created, we also ensured that the function the user provided is valid.

> This was a simple solution we tried initially and it worked well - until users wanted more capabilities. One capability was **Value Compute Functions** - functions that returned a vector of values. Users wanted to select a value from a dynamic list of potential values that changed often which was fetched from internal systems at Juspay. Another downside to our nodeJS isolate solution was a lot of wrapper code to ensure that we properly isolated user code. We didn't validate if the code that was being stored was valid syntatically, and if node could not execute it we didn't return that error to the user.

Reviewer 2: just internal systems is sufficient - need not refer to company name unless absolutely needed.

Author: Any reason to not mention Juspay? Would help contextualize `internal systems` and helps readers understand that this is used somewhere

> This was a simple solution we tried initially and it worked well - until users wanted more capabilities. One capability was **Value Compute Functions** - functions that returned a vector of values. Users wanted to select a value from a dynamic list of potential values that changed often which was fetched from internal systems at Juspay. Another downside to our nodeJS isolate solution was a lot of wrapper code to ensure that we properly isolated user code. We didn't validate if the code that was being stored was valid syntatically, and if node could not execute it we didn't return that error to the user.

Reviewer 2: - `that changed often which was fetched …` — subject-verb confusion (`values … was`) and a missing comma before `which`.
- `syntatically` → `syntactically`.
- `if node could not execute it we didn't return …` — missing comma after `it`; capitalise `Node`.

```suggestion
This was a simple solution we tried initially and it worked well - until users wanted more capabilities. One capability was **Value Compute Functions** - functions that returned a vector of values. Users wanted to select a value from a dynamic list of values, often fetched from internal systems at Juspay, that changes frequently. Another downside to our Node.js isolate solution was a lot of wrapper code to ensure that we properly isolated user code. We didn't validate whether the stored code was syntactically valid, and if Node could not execute it, we didn't return that error to the user.
```

# Other options

> The team had previously looked into [`mlua`](https://crates.io/crates/mlua) and looked at ways to run javascript from within rust to have better interoperability. Two options we considered were [`rquickjs`](https://crates.io/crates/rquickjs) and [`rustyscript`](https://crates.io/crates/rustyscript); we went with rustyscript because of its [documentation](https://rscarson.github.io/rustyscript-book/), features and closer compatibility with nodeJS.

Reviewer 2: - Missing Oxford comma (`documentation, features, and …`).
- `closer compatibility with nodeJS` — closer than what? Either say `closer to Node.js than rquickjs` or just `better Node.js compatibility`.
- It would also be useful to readers if you said *why* `mlua` was dropped and what specifically you compared between `rquickjs` and `rustyscript` (perf? embedding ergonomics? Deno API surface?).

```suggestion
The team had previously looked into [`mlua`](https://crates.io/crates/mlua) and explored ways to run JavaScript from within Rust to get better interoperability. Two options we considered were [`rquickjs`](https://crates.io/crates/rquickjs) and [`rustyscript`](https://crates.io/crates/rustyscript); we went with rustyscript because of its [documentation](https://rscarson.github.io/rustyscript-book/), features, and better Node.js compatibility.
```

# Running Javascript in Rust with Rustyscript

> Rustyscript is powerful and extremely customizable for running user code - it has a permissions model, functions that you can expose from rust to use in JS land and some very nice built in functionality like console, crypto, http, etc. 

Reviewer 2: ```suggestion
Rustyscript is powerful and extremely customizable for running user submitted JS code - it has a permissions model, functions that you can expose from rust to use in JS land and some very nice built in functionality like console, crypto, http, etc. 
```

Author: The title mentions javascript so I don't want to repeat it too many times and the word `submitted` is implied

> Rustyscript is powerful and extremely customizable for running user code - it has a permissions model, functions that you can expose from rust to use in JS land and some very nice built in functionality like console, crypto, http, etc. 

Reviewer 2: `rust` → `Rust`, `built in` → `built-in`, missing Oxford comma, trailing whitespace.

```suggestion
Rustyscript is powerful and extremely customizable for running user code - it has a permissions model, functions that you can expose from Rust to use in JS land, and some very nice built-in functionality like console, crypto, http, etc.
```

> Here's a function to valid if a link is valid by calling it and checking if an `OK` response code is sent back:

Reviewer 2: Typo (`valid` → `validate`) and the duplicated `valid … valid` reads badly.

```suggestion
Here's a function that validates a URL by fetching it and checking that the response status is 200:
```

```javascript
async function execute(payload) {
    const { value_validate } = payload;
    // value_validate is an object that contains all arguments a validation function would need
    const { key, value, type, environment } = value_validate; 
    let response = await fetch(value);

>     return response.status == '200';

Reviewer: `return response.status == 200;`

Author: Both work

>     return response.status == '200';

Reviewer 2: `response.status` is a number; comparing it to the string `'200'` works only because of `==` coercion. Use `response.ok` (or `=== 200`) so the example doesn't model a subtle anti-pattern.

```suggestion
    return response.ok;
```

Author: We need to change a few functions internally then

}
```

> All functions in Superposition are named `execute`. This is because during validating the user input, we wrap this code with some helpers and the final code that is passed to rustyscript looks like:

Reviewer 2: `during validating` is ungrammatical.

```suggestion
All functions in Superposition are named `execute`. This is because, while validating the user input, we wrap this code with some helpers and the final code that is passed to rustyscript looks like:
```

```javascript
async function execute(payload) {
    const { value_validate } = payload;
    // value_validate is an object that contains all arguments a validation function would need
    const { key, value, type, environment } = value_validate; 
    let response = await fetch(value);
    return response.status == '200';
}
export function typeCheck() {
    if (typeof execute === "undefined") {
        throw new Error("execute function is not defined");
    }

> }

Reviewer: export function typeCheck() {
    if (typeof execute !== "function") {
        throw new Error("execute must be a function");
    }
}

Author: This is what is in superposition

Reviewer 2: We should then fix it in superposition then ideally.

```

> We validate this with `rustyscript::validate`, which takes a string and tells you if its valid JS syntax:

Reviewer 2: `its` → `it's`.

```suggestion
We validate this with `rustyscript::validate`, which takes a string and tells you if it's valid JS syntax:
```

```rust
rustyscript::validate(&type_check_code).map_err(|err| {
    log::error!("Invalid function syntax: {:?}", err);
    bad_argument!("Invalid function syntax: {}", err)
})?;
```
This validates the syntax, but is `execute` a function? To run a function in rustyscript we need to:

- Create a basic runtime
- Load a javascript module,

> - Call a function and the resulting value

Reviewer 2: This point seems incomplete - did you mean and 

> consume the resulting value ?

> - Call a function and the resulting value

Reviewer 2: List items aren't parallel: L96 ends with a comma, L97 is missing a verb (`Call a function and the resulting value`). Also `javascript` → `JavaScript`.

```suggestion
- Create a basic runtime
- Load a JavaScript module
- Call a function and get the resulting value
```

We check `execute` by running and calling `typeCheck`:

```rust
// create a basic runtime
let module = Module::new("type_check.js", &type_check_code);
let runtime_options = RuntimeOptions {
    timeout: Duration::from_millis(1500),
    ..Default::default()
};

let mut runtime = Runtime::new(runtime_options).map_err(|e| {
    let err_str = e.to_string();
    validation_error!("Failed to create runtime: {}", err_str)
})?;

let module_handle = runtime.load_module(&module).map_err(|e| {
    let err_str = e.to_string();
    validation_error!("Failed to load module: {}", err_str)
})?;
// typeCheck is not an async function, but call_function_async is used as a 
// catch all for sync and async code
let tokio_runtime = runtime.tokio_runtime();
tokio_runtime
    .block_on(async {
        runtime
            .call_function_async::<()>(
                Some(&module_handle),
                "typeCheck",
                json_args!(),
            )
            .await
    })
    .map_err(|e| {
        let err_str = e.to_string();
        validation_error!("Function validation failed: {}", err_str)
    })?;

Ok(())
```

> Now when this function needs to be run to validate any input or change, it is wrapped in the below JS code. Our wrapper code for user functions in rustyscript looks like this, we override console so that it can output logs to an array:

Reviewer 2: Comma splice (`looks like this, we override …`) and the sentence is doing two things at once. Suggest splitting and leading with what the wrapper actually does so the reader knows what to look for in the snippet.

```suggestion
Now when this function needs to be run to validate any input or change, it is wrapped in the JS code below. Note the console override — it routes every log call into an in-memory buffer that we expose via `getLogBuffer`:
```

    timeout: Duration::from_millis(1500),
    ..Default::default()
};

let mut runtime = Runtime::new(runtime_options).map_err(|e| {
    let err_str = e.to_string();
    (format!("Failed to create runtime: {}", err_str), None)
})?;

let payload = FunctionPayload {
    version: runtime_version,
    payload: args.clone(),
};
let module_handle = runtime
    .load_module(&module)
    .map_err(|err| (err.to_string(), None))?;

let tokio_runtime = runtime.tokio_runtime();
// execute function, fn_output takes the value of Value::Bool for the example
let fn_output = tokio_runtime
    .block_on(async {
        runtime
            .call_function_async::<serde_json::Value>(
                Some(&module_handle),
                "execute",
                json_args!(payload),
            )
            .await
    })
    .map_err(|err| (err.to_string(), None))?;

// get logs
let stdout = runtime
    .call_function::<Vec<String>>(Some(&module_handle), "getLogBuffer", json_args!())
    .map_err(|err| (err.to_string(), None))?
    .join("\n");

runtime
    .call_function::<()>(Some(&module_handle), "clearLogBuffer", json_args!())
    .map_err(|err| (err.to_string(), None))?;
```

# Conclusion

> Rustyscript can do a lot more than what we've shared here - I recommend taking a look at the [documentation](https://rscarson.github.io/rustyscript-book/index.html) if you've never used it before. For Superposition, its ability to run isolated javascript seamlessly from within rust has enabled our team to provide more useful features to customers.

Reviewer 2: The conclusion is thin and doesn't tie back to the opening pain point (correctness beyond types). Quantifying/qualifying the win lands harder.

```suggestion
Rustyscript can do a lot more than what we've shared here - I recommend taking a look at the [documentation](https://rscarson.github.io/rustyscript-book/index.html) if you've never used it before. For Superposition, Rustyscript's ability to run isolated JavaScript seamlessly from within Rust has unlocked richer correctness guarantees — we now support both validation and value-compute functions with full syntax checking and captured logs, without spawning a subprocess.
```

> If you like the idea of Superposition, do take a look at our [docs](https://juspay.io/superposition/docs) to learn more. Have questions? Join our [slack](superpositionjp.slack.com) to interact with the team.

Reviewer: Slack link is broken. Can you fix it ?

Author: Broken how? This is what we have in readme

> If you like the idea of Superposition, do take a look at our [docs](https://juspay.io/superposition/docs) to learn more. Have questions? Join our [slack](superpositionjp.slack.com) to interact with the team.

Reviewer 2: `[slack](superpositionjp.slack.com)` — link is missing the `https://` scheme so it renders as a relative path; also capitalise `Slack`.

```suggestion
If you like the idea of Superposition, do take a look at our [docs](https://juspay.io/superposition/docs) to learn more. Have questions? Join our [Slack](https://superpositionjp.slack.com) to interact with the team.
```

---
slug: js-sandbox-in-rust
title: Rust JS Sandbox - Running Untrusted JavaScript Inside Superposition

> description: Designing and implementing validation and value-compute functions: how we run user-defined JS inside Rust, seamlessly, and in isolation.

Reviewer 3: Quote the values, getting 
`Error in user YAML: (<unknown>): mapping values are not allowed in this context at line 3 column 79`
maybe because of the : in the description

---

TL;DR - We use Rustyscript to sandbox and execute user-submitted JS code inside our Rust application

[Superposition](https://github.com/juspay/superposition) is a context-aware configuration management system that ensures correct, safe, and reliable config changes. One of the ways we've achieved this goal is by using jsonschema types for all configuration values with a default value. We call this concept [Default Configs](/docs/basic-concepts/context-aware-config/default-config). We soon realized an issue - though the type is right, that does not guarantee the correctness of the actual value.

# Where types aren't enough

Take the following configuration:

```json
{
    "image_url": "https://example.com/image/cat.png"
}
```

The strictest JSON schema for the value of image_url would be:
```json
{
  "type": "string",
  "pattern": "^https?:\\/\\/(www\\.)?[-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6}\\b([-a-zA-Z0-9()@:%_\\+.~#?&//=]*)$"
}
```
This schema validates the type and shape and that is one part of correctness, but we're missing the most critical aspect - does the URL exist? Does it return what we're expecting?

# User Defined Validation Functions!

> We introduced *Functions* - a way for any end user to write validation logic and link them to a Default Config or [Dimension](/docs/basic-concepts/context-aware-config/dimensions) to ensure the correctness of that value (should they want to do that). These functions are written in our frontend monaco editor and saved to a database with some other information. The function looks like this:

Reviewer 2: link to monaco ?  generally a nice thing to link to other projects.

When a function is being edited/created, we also ensured that the function the user provided is valid.

This was a simple solution we tried initially and it worked well - until users wanted more capabilities. One capability was **Value Compute Functions** - functions that returned a vector of values. Users wanted to select a value from a dynamic list of potential values, fetched from internal systems at Juspay, that changes frequently. Another downside to our Node.js isolate solution was a lot of wrapper code to ensure that we properly isolated user code. We didn't validate if the code that was being stored was valid syntactically, and if Node.js could not execute it, we didn't return that error to the user.

# Other options

> The team had previously looked into [`mlua`](https://crates.io/crates/mlua)(which we decided to not pursue since Superposition users were more familiar with Javascript and looked at ways to run Javascript from within Rust to have better interoperability. Two options we considered were [`rquickjs`](https://crates.io/crates/rquickjs) and [`Rustyscript`](https://crates.io/crates/rustyscript); we went with Rustyscript because of its [documentation](https://rscarson.github.io/rustyscript-book/), features, and better compatibility with Node.js.

Reviewer 3: nitpick: `(` not being closed.

    }
};

Object.defineProperty(globalThis, 'console', {
    value: customConsole,
    writable: false,
    configurable: false
});

function getLogBuffer() {
    return logBuffer;
}

function clearLogBuffer() {
    logBuffer = [];
}

// actual user function
async function execute(payload) {
    const { value_validate } = payload;
    // value_validate is an object that contains all arguments a validation function would need
    const { key, value, type, environment } = value_validate; 
    let response = await fetch(value);
    return response.status === 200;
}

export { execute, getLogBuffer, clearLogBuffer };
```

Note the console override — it routes every log call into an in-memory buffer that we expose via `getLogBuffer`.

We validate the syntax of the above Javascript with `rustyscript::validate`, which takes a string and tells you if it is valid syntax or not. If it is not valid, we return an error to the user. This is done before we even try to run the function:

```rust
rustyscript::validate(&type_check_code).map_err(|err| {
    log::error!("Invalid function syntax: {:?}", err);
    bad_argument!("Invalid function syntax: {}", err)
})?;
```

> This validates the syntax, but is `execute` a functio and does it return a value that Superposition expects from this category of function? With rustyscript, we can simply just run the function with some default inputs and inspect its return value. This checks:

Reviewer 3: ```suggestion
This validates the syntax, but is `execute` a function and does it return a value that Superposition expects from this category of function? With rustyscript, we can simply just run the function with some default inputs and inspect its return value. This checks:
```
